### DISCUSSION TITLE
Route and service types

### DISCUSSION THREAD

**Discussion by kellytk** [tokio-rs/axum#346] [Q&A] [answered]
I can use a service inline:

`.route("/Cargo.toml", service::get(ServeFile::new("Cargo.toml")))`

And separately if I have a handler:

```
async fn handle() -> String {
    String::from("handle")
}
```
I can use it:

`.route("/", handler::get(handle))`

How can the same be done with the service, moving its body into a separate function?

```
async fn file_service() -> ServeFile {
    ServeFile::new("Cargo.toml")
}
```

`.route("/Cargo.toml", service::get(file_service))`

Generates the error:

```
error[E0277]: the trait bound `fn() -> impl Future {file_service}: tower_service::Service<Request<axum::body::Body>>` is not satisfied
  --> src/main.rs:11:31
   |
11 |         .route("/Cargo.toml", service::get(file_service));
   |                               ^^^^^^^^^^^^ the trait `tower_service::Service<Request<axum::body::Body>>` is not implemented for `fn() -> impl Future {file_service}`
```

**Comment by davidpdrsn** [comment, marked as answer]
You can make a handler that calls a service and returns it's response. The most common way is `some_service.oneshot(request).await` but there are a number useful methods on [`tower::ServiceExt`](https://docs.rs/tower/0.4.8/tower/trait.ServiceExt.html#method.oneshot).

**Reply by kellytk** [reply]
Thank you.  How should the return value of `some_service.oneshot(request).await` be converted to a value acceptable to `handler::get()`?

`.route("/Cargo.toml", handler::get(file_handler))`

```
async fn file_handler(
    request: Request<Body>,
) -> Result<Response<ServeFileResponseBody>, std::io::Error> {
    ServeFile::new("Cargo.toml")
        .oneshot(request)
        .await
}
```

Generates the error:

```
error[E0277]: the trait bound `fn(Request<axum::body::Body>) -> impl Future {file_handler}: Handler<_, _>` is not satisfied
   --> src/main.rs:13:44
    |
13  |         .route("/Cargo.toml", handler::get(file_handler));
    |                                            ^^^^^^^^^^^^ the trait `Handler<_, _>` is not implemented for `fn(Request<axum::body::Body>) -> impl Future {file_handler}`
```

**Reply by davidpdrsn** [reply]
As mentioned [in the docs](https://docs.rs/axum/0.2.5/axum/#building-responses) handlers must return something that implements `IntoResponse` which `std::io::Error` does not. So you have to somehow convert that into a response that's appropriate for your app. You can `match` on the result or use `Result::map_err`.

**Reply by kellytk** [reply]
By [temporarily] adding `unwrap` to `oneshot`'s `await` and stripping the `Result` and `Error` from the return type the code now compiles.  Thanks again.
